### PR TITLE
nixos: opengl -> impure-drivers.opengl

### PR DESCRIPTION
**Pull request by SomeoneSerge**
## Description of changes

WIP. Following https://github.com/NixOS/nixpkgs/pull/275241 and https://github.com/NixOS/nixpkgs/pull/269475

The motivation is that it's extremely confusing for the new users when instructions for enabling something like CUDA involve setting `hardware.opengl.enable` and `services.xserver.videoDrivers`. The module's new name is meant to suggest that it's there to ensure integration with the `addDriverRunpath`, and it hopefully explains why/when the module has to be enabled (when we're forced to link drivers "impurely").

I think it'd also make sense to introduce an option like `hardware.impure-drivers.cuda.enable`, which would guide the user through the process of setting up the x11 or datacenter drivers (e.g. through assertions). I'd keep that for a separate PR

Thoughts? @jonringer @Kiskae @NixOS/cuda-maintainers @NixOS/rocm-maintainers

## Things done

- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
  - [x] `nixosTests.quake3`
  - [x] `nixosTests.tinywl`
  - [ ] `nixosTests.openarena`
  - [ ] `nixosTests.seatd`
  - [ ] `nixosTests.cagebreak`
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/hardware/opengl.nix [hunk 1]
```diff
@@ -129,32 +156,35 @@ in
...
 
     systemd.tmpfiles.rules = [
-      "L+ /run/opengl-driver - - - - ${package}"
+      "L+ ${impurePath} - - - - ${package}"
       (
         if pkgs.stdenv.isi686 then
-          "L+ /run/opengl-driver-32 - - - - opengl-driver"
-        else if cfg.driSupport32Bit then
-          "L+ /run/opengl-driver-32 - - - - ${package32}"
+          "L+ ${impurePath32} - - - - ${impurePath}"
```
Comment (SomeoneSerge) [inline nixos/modules/hardware/opengl.nix:172]: The target used to be a relative path (`"opengl-driver"`), now it's absolute (`"/run/opengl-driver"`). Shouldn't matter

File: nixos/modules/hardware/opengl.nix [hunk 2]
```diff
@@ -4,35 +4,90 @@ with lib;
...
 
   imports = [
-    (mkRenamedOptionModule [ "services" "xserver" "vaapiDrivers" ] [ "hardware" "opengl" "extraPackages" ])
+    (mkRenamedOptionModule [ "services" "xserver" "vaapiDrivers" ] [ "hardware" "impure-drivers" "packages" ])
     (mkRemovedOptionModule [ "hardware" "opengl" "s3tcSupport" ] "S3TC support is now always enabled in Mesa.")
+
+    # The old option `hardware.opengl.enable` corresponds to `hardware.impure-drivers.opengl.enable` (defaults to `true`).
+    # However, in order support transition of the systems have been setting `hardware.opengl.enable`,
+    # we need to translate it into `hardware.impure-drivers.enable` (defaults to `false`)
+    (mkRenamedOptionModule [ "hardware" "opengl" "enable" ] [ "hardware" "impure-drivers" "enable" ])
```
Comment (SomeoneSerge) [inline nixos/modules/hardware/opengl.nix:37]: I think the most questionable decision is this. Maybe I shouldn't be introducing `hardware.impure-drivers.opengl.enable` at all?

Comment (Kiskae) [reply nixos/modules/hardware/opengl.nix]: Perhaps leave `hardware.opengl.enable` unchanged and make `hardware.impure-drivers.opengl.enable` depend on its value?

Comment (SomeoneSerge) [reply nixos/modules/hardware/opengl.nix]: H'm, how about I keep `hardware.opengl.enable`, drop `hardware.impure-drivers.opengl`, but add `hardware.impure-drivers.enable = true` to `hardware.opengl`'s `config`?

**Comment by Tungsten842** [comment]
There is another attempt at this, not sure about the status: https://github.com/NixOS/nixpkgs/pull/158079.
I personally prefer hardware.drivers compared to hardware.impure-drivers.

**Comment by SomeoneSerge** [comment]
> There is another attempt at this, not sure about the status: https://github.com/NixOS/nixpkgs/pull/158079.
> I personally prefer hardware.drivers compared to hardware.impure-drivers.

Oooh right, I forgot about those

**Comment by Atemu** [comment]
I haven't had much time to work on it and that likely won't change until April.

The main point I've been stuck on is that I wanted to build a better pattern to select packages that require multiple bitnesses. For example, right now we have this ugly `package` `package32` pattern where you must set two options to change anything and it's possible to set those to different things (other than bitness) or forget to set one.

I've tried a few different things and for now the best I've come up with is building a "package selector" type which is a function from `attrset -> drv` that takes `pkgs` as an argument and returns a single package. In the implementation, you'd simply call it with the relevant package set.

Again, I haven't spent as much time on this as I had hoped.

**Comment by jonringer** [comment]
I've been hopign to do the refactor as well. But wanted to record it to demonstrate how to deprecate old options. just haven't had a single window to free time to do so.

**Pull request closed without merging**